spring-boot-project/spring-boot-actuator-autoconfigure/src/test/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/DualPrometheusMetricsExportAutoConfigurationTests.java

```diff
@@ -0,0 +1,409 @@
+				.hasSingleBean(io.micrometer.prometheus.PrometheusConfig.class)
+				.hasBean("customConfig")
+				.hasSingleBean(PrometheusRegistry.class)
+				.hasSingleBean(io.micrometer.prometheusmetrics.PrometheusConfig.class)
+				.hasBean("otherCustomConfig"));
+	}
+
+	@Test
+	void allowsCustomRegistryToBeUsed() {
+		this.contextRunner.withUserConfiguration(CustomRegistryConfiguration.class)
+			.run((context) -> assertThat(context).hasSingleBean(io.micrometer.prometheus.PrometheusMeterRegistry.class)
+				.hasBean("customRegistry")
+				.hasSingleBean(CollectorRegistry.class)
+				.hasSingleBean(io.micrometer.prometheus.PrometheusConfig.class)
+				.hasSingleBean(PrometheusRegistry.class)
+				.hasSingleBean(io.micrometer.prometheusmetrics.PrometheusConfig.class)
+				.hasBean("otherCustomRegistry"));
+	}
+
+	@Test
+	void allowsCustomCollectorRegistryToBeUsed() {
+		this.contextRunner.withUserConfiguration(CustomCollectorRegistryConfiguration.class)
+			.run((context) -> assertThat(context).hasSingleBean(io.micrometer.prometheus.PrometheusMeterRegistry.class)
+				.hasBean("customCollectorRegistry")
+				.hasSingleBean(CollectorRegistry.class)
+				.hasSingleBean(io.micrometer.prometheus.PrometheusConfig.class)
+				.hasBean("customPrometheusRegistry")
+				.hasSingleBean(PrometheusRegistry.class)
+				.hasSingleBean(io.micrometer.prometheusmetrics.PrometheusConfig.class));
+	}
+
+	@Test
+	void autoConfiguresExemplarSamplerIfSpanContextSupplierIsPresent() {
+		this.contextRunner.withUserConfiguration(ExemplarsConfiguration.class)
+			.run((context) -> assertThat(context).hasSingleBean(SpanContextSupplier.class)
+				.hasSingleBean(ExemplarSampler.class)
+				.hasSingleBean(io.micrometer.prometheus.PrometheusMeterRegistry.class));
+	}
+
+	@Test
+	void allowsCustomExemplarSamplerToBeUsed() {
+		this.contextRunner.withUserConfiguration(ExemplarsConfiguration.class)
+			.withBean("customExemplarSampler", ExemplarSampler.class, () -> mock(ExemplarSampler.class))
+			.run((context) -> assertThat(context).hasSingleBean(ExemplarSampler.class)
+				.getBean(ExemplarSampler.class)
+				.isSameAs(context.getBean("customExemplarSampler")));
+	}
+
+	@Test
+	void exemplarSamplerIsNotAutoConfiguredIfSpanContextSupplierIsMissing() {
+		this.contextRunner.withUserConfiguration(BaseConfiguration.class)
+			.run((context) -> assertThat(context).doesNotHaveBean(SpanContextSupplier.class)
+				.doesNotHaveBean(ExemplarSampler.class)
+				.hasSingleBean(io.micrometer.prometheus.PrometheusMeterRegistry.class));
+	}
+
+	@Test
+	void addsScrapeEndpointToManagementContext() {
+		this.contextRunner.withConfiguration(AutoConfigurations.of(ManagementContextAutoConfiguration.class))
+			.withUserConfiguration(BaseConfiguration.class)
+			.withPropertyValues("management.endpoints.web.exposure.include=prometheus")
+			.run((context) -> assertThat(context).hasSingleBean(PrometheusScrapeEndpoint.class)
+				.doesNotHaveBean(PrometheusSimpleclientScrapeEndpoint.class));
```

Comment:
This test class is to demonstrate the case when both `io.micrometer:micrometer-registry-prometheus` and `io.micrometer:micrometer-registry-prometheus-simpleclient` are on the classpath. Mostly beans from both will be created, but there should only be one scrape endpoint and the one using `io.micrometer:micrometer-registry-prometheus` takes precedence.